src/components/SetupTab/index.jsx

```diff
@@ -2,41 +2,354 @@ import React from 'react';
+  let argsValidation = {};
+  let argsValues = {};
+  Object.keys(argsSpec).forEach((argkey) => {
+    argsValidation[argkey] = {}
+    if (argkey === 'n_workers') { return }
+    argsValues[argkey] = {
+      value: argsDict[argkey],
+      touched: !initIsEmpty  // touch them only if initializing with values
+    }
+  });
+  return({ argsValues: argsValues, argsValidation: argsValidation })
+}
 
 export class SetupTab extends React.Component {
-  /** Renders an Arguments form and an Execute button
+  /** Renders an arguments form, execute button, and save button.
   */
+  constructor(props) {
+    super(props);
+    this.state = {
+      argsValues: null,
+      argsValidation: {},
+      argsValid: false,
+      sortedArgTree: null
+    }
+
+    this.savePythonScript = this.savePythonScript.bind(this);
+    this.wrapArgsToJsonFile = this.wrapArgsToJsonFile.bind(this);
+    this.wrapInvestExecute = this.wrapInvestExecute.bind(this);
+    this.investValidate = this.investValidate.bind(this);
+    this.updateArgValues = this.updateArgValues.bind(this);
+    this.batchUpdateArgs = this.batchUpdateArgs.bind(this);
+  }
+
+  componentDidMount() {
+    /* 
+    * Including the `key` property on SetupTab tells React to 
+    * re-mount (rather than update & re-render) this component when
+    * the key changes. This function does some useful initialization
+    * that only needs to compute when this.props.argsSpec changes,
+    * not on every re-render.
+    */
+    if (this.props.argsInitValues) {
+      let argTree = {}
+      let { argsValues, argsValidation } = initializeArgValues(
+        this.props.argsSpec, this.props.argsInitValues)
+      
+      Object.keys(this.props.argsSpec).forEach((argkey) => {
+        if (argkey === 'n_workers' || 
+          this.props.argsSpec[argkey].order === 'hidden') { return }
+        const argSpec = Object.assign({}, this.props.argsSpec[argkey])
+        if (argSpec.ui_control) {
+          argsValues = toggleDependentInputs(this.props.argsSpec, argsValues, argkey)
+        }
+
+        // This grouping and sorting does not fail if order is undefined 
+        // (i.e. it is missing from the UI spec) for one or more args. 
+        // Nevertheless, feels better to fill in a float here.
+        if (typeof argSpec.order !== 'number') { argSpec.order = 100.0 }
+
+        // Fill in a tree-like object where each item is an array of the args
+        // that share a major (Math.floor) order number and should be grouped.
+        // Within each array representing a group, items are objects like:
+        //  { <precise order number> : <argkey> }
+        const group = Math.floor(argSpec.order)
+        let subArg = {}
+        if (argTree[group]) {  // already have arg(s) in this group
+          subArg[argSpec.order] = argkey
+          argTree[group].push(subArg)
+        } else {
+          subArg[argSpec.order] = argkey
+          argTree[group] = [subArg]
+        }
+      });

```

Comment:
Ah, very interesting grouping technique!

Do we have any 3rd-order groupings to worry about?